client/client.py

```diff
@@ -0,0 +1,135 @@
+import pickle
+import socket
+import threading
+
+from numpy import full
+
+from notebook.notebook import DistributedNotebook
+
+RECV_BUFFER = 1024
+
+class NotebookClient():
+    """
+    NotebookClient handles syncing with remote peers to implement asynchronous
+    collaboration.
+    """
+
+    def __init__(self, port, peers, name="alice"):
+        self.port = int(port)
+        self.peers = {}
+        for peer in peers:
+            peer_name, peer_port = peer.split(":")
+            self.peers[peer_name] = peer_port
+        self.name = name
+        self.notebook = DistributedNotebook(id=port)
+        self.lock = threading.Lock()
+        self.editor = None
+
+    def attach_editor(self, editor):
+        """
+        Attaches a NotebookEditor to the client to enable editor updates from listen().
+        """
+        self.editor = editor
+
+    def get_peers(self):
+        """
+        Returns the list of peer names.
+        """
+        return list(self.peers.keys())
+    
+    def host(self):
+        """
+        Starts a listener thread to receive sync messages from remote peers.
+        """
+        listen = threading.Thread(target=self.listen, args=(self.port,))
+        listen.start()
+
+    def send_bytes(self, sock, data):
+        """
+        Send data over a socket.
+        """
+        size = len(data).to_bytes(4, byteorder='big')
+        sock.sendall(size)
+        sock.sendall(data)
+
+    def recv_bytes(self, sock):
+        """
+        Receive data from a socket.
+        """
+        size = sock.recv(4)
+        size = int.from_bytes(size, byteorder='big')
+        data = b''
+        while len(data) < size:
+            buffer = sock.recv(size - len(data))
+            if not buffer:
+                raise EOFError("Connection closed")
+            data += buffer
+        return data
+
+    def listen(self, port):
+        """
+        Listens for sync messages from remote peers.
+        """
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(('localhost', port))
```

Comment:
What do you think of making `host` a user-provided param on init (defaulting to `localhost`), or perhaps a global variable defined at the top of the file? Do you think it would help with presenting a pathway to using something like this in a real distributed context?